crates/bevy_ui/src/layout/mod.rs

```diff
@@ -241,64 +244,110 @@ pub enum LayoutError {
 #[allow(clippy::too_many_arguments)]
 pub fn ui_layout_system(
     primary_window: Query<(Entity, &Window), With<PrimaryWindow>>,
-    windows: Query<(Entity, &Window)>,
+    cameras: Query<(Entity, &Camera)>,
     ui_scale: Res<UiScale>,
     mut scale_factor_events: EventReader<WindowScaleFactorChanged>,
     mut resize_events: EventReader<bevy_window::WindowResized>,
     mut ui_surface: ResMut<UiSurface>,
-    root_node_query: Query<Entity, (With<Node>, Without<Parent>)>,
-    style_query: Query<(Entity, Ref<Style>), With<Node>>,
+    root_node_query: Query<(Entity, Option<&TargetCamera>), (With<Node>, Without<Parent>)>,
+    style_query: Query<(Entity, Ref<Style>, Option<&TargetCamera>), With<Node>>,
     mut measure_query: Query<(Entity, &mut ContentSize)>,
     children_query: Query<(Entity, Ref<Children>), With<Node>>,
     just_children_query: Query<&Children>,
     mut removed_children: RemovedComponents<Children>,
     mut removed_content_sizes: RemovedComponents<ContentSize>,
-    mut node_transform_query: Query<(&mut Node, &mut Transform)>,
     mut removed_nodes: RemovedComponents<Node>,
+    mut node_transform_query: Query<(&mut Node, &mut Transform)>,
 ) {
-    // assume one window for time being...
-    // TODO: Support window-independent scaling: https://github.com/bevyengine/bevy/issues/5621
-    let (primary_window_entity, logical_to_physical_factor, physical_size) =
-        if let Ok((entity, primary_window)) = primary_window.get_single() {
-            (
-                entity,
-                primary_window.resolution.scale_factor(),
-                Vec2::new(
-                    primary_window.resolution.physical_width() as f32,
-                    primary_window.resolution.physical_height() as f32,
-                ),
-            )
-        } else {
-            return;
-        };
-
-    let resized = resize_events
-        .read()
-        .any(|resized_window| resized_window.window == primary_window_entity);
-
-    let scale_factor = logical_to_physical_factor * ui_scale.0;
-
-    let layout_context = LayoutContext::new(scale_factor, physical_size);
-
-    if !scale_factor_events.is_empty() || ui_scale.is_changed() || resized {
-        scale_factor_events.clear();
-        // update all nodes
-        for (entity, style) in style_query.iter() {
-            ui_surface.upsert_node(entity, &style, &layout_context);
+    struct CameraLayoutInfo {
+        size: UVec2,
+        resized: bool,
+        scale_factor: f32,
+        root_nodes: Vec<Entity>,
+    }
+
+    // If there is only one camera, we use it as default
+    let default_single_camera = cameras.get_single().ok().map(|(entity, _)| entity);
```

Comment:
i wonder if it's better to filter cameras by those with RenderTarget::Window to find the default?

i added 
```rust
#[derive(SystemParam)]
pub struct DefaultUiCamera<'w, 's> {
    cameras: Query<'w, 's, (Entity, &'static Camera)>,
}

impl<'w, 's> DefaultUiCamera<'w, 's> {
    pub fn get(&self) -> Option<Entity> {
        let mut window_cameras = self.cameras
            .iter()
            .filter(|(_, c)| matches!(c.target, RenderTarget::Window(_)));
        match (window_cameras.next(), window_cameras.next()) {
            (first, None) => first.map(|(entity, _)| entity),
            _ => None,
        }
    }
}
```

and then used it everywhere to set `default_single_camera` (using `Extract<DefaultUiCamera>` in the extract systems), it works fine and has the (maybe minimal) advantage that existing render-to-texture scenarios will continue to work as before.

Reply:
it's probably better still to use `With<PrimaryWindow>`

Reply:
apologies for the noise, i ended up finally with

```rust
#[derive(SystemParam)]
pub struct DefaultUiCamera<'w, 's> {
    cameras: Query<'w, 's, (Entity, &'static Camera)>,
    primary_window: Query<'w, 's, Entity, With<PrimaryWindow>>,
}

impl<'w, 's> DefaultUiCamera<'w, 's> {
    pub fn get(&self) -> Option<Entity> {
        self.cameras
            .iter()
            .filter(|(_, c)| match c.target {
                RenderTarget::Window(WindowRef::Primary) => true,
                RenderTarget::Window(WindowRef::Entity(w)) => self.primary_window.get(w).is_ok(),
                _ => false,
            })
            .max_by_key(|(e, c)| (c.order, *e))
            .map(|(e, _)| e)
    }
}
```

this chooses the latest-ordered camera pointing at the primary window, using entity id to disambiguate if there are multiple cameras with the same order (which triggers a warning elsewhere already).

Reply:
Done this, thanks for the contribution!